File: library/src/main/java/name/zeno/android/behavior/HideBehavior.java
package name.zeno.android.behavior;

import android.content.Context;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.view.ViewCompat;
import android.support.v4.view.ViewPropertyAnimatorCompat;
import android.support.v4.view.animation.LinearOutSlowInInterpolator;
import android.util.AttributeSet;
import android.view.View;

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 16/8/17
 */
@SuppressWarnings("unused")
public class HideBehavior extends VerticalScrollingBehavior
{
  private ViewPropertyAnimatorCompat anim;
  private boolean hide = false;

  public HideBehavior(Context context, AttributeSet attrs)
  {
    super(context, attrs);
  }

  @Override
  public void onDirectionNestedPreScroll(CoordinatorLayout coordinatorLayout, View child, View target, int dx, int dy, int[] consumed, @ScrollDirection int scrollDirection)
  {
    showOrHideView(child, target, scrollDirection);
  }

  @Override
  protected boolean onNestedDirectionFling(CoordinatorLayout coordinatorLayout, View child, View target, float velocityX, float velocityY, @ScrollDirection int scrollDirection)
  {
    showOrHideView(child, target, scrollDirection);
    return true;
  }

  private void showOrHideView(View child, View target, @ScrollDirection int direction)
  {
    boolean willHide = direction == ScrollDirection.SCROLL_DIRECTION_UP;
    if (this.hide == willHide) return;

    this.hide = willHide;

    if (anim == null) {
      anim = ViewCompat.animate(child);
      anim.setDuration(300);
      anim.setInterpolator(new LinearOutSlowInInterpolator());
    } else {
      anim.cancel();
    }
    anim.scaleX(direction == ScrollDirection.SCROLL_DIRECTION_UP ? 0 : 1)
        .scaleY(direction == ScrollDirection.SCROLL_DIRECTION_UP ? 0 : 1)
        .start();
  }

}

File: library/src/main/java/name/zeno/android/behavior/HideBehavior.kt
package name.zeno.android.behavior

import android.content.Context
import android.support.design.widget.CoordinatorLayout
import android.support.v4.view.ViewCompat
import android.support.v4.view.ViewPropertyAnimatorCompat
import android.support.v4.view.animation.LinearOutSlowInInterpolator
import android.util.AttributeSet
import android.view.View

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since 16/8/17
 */
class HideBehavior(context: Context, attrs: AttributeSet) : VerticalScrollingBehavior<View>(context, attrs) {
  private var anim: ViewPropertyAnimatorCompat? = null
  private var hide = false

  override fun onDirectionNestedPreScroll(coordinatorLayout: CoordinatorLayout, child: View, target: View, dx: Int, dy: Int, consumed: IntArray, @ScrollDirection scrollDirection: Int) {
    showOrHideView(child, scrollDirection)
  }

  override fun onNestedDirectionFling(coordinatorLayout: CoordinatorLayout, child: View, target: View, velocityX: Float, velocityY: Float, @ScrollDirection scrollDirection: Int): Boolean {
    showOrHideView(child, scrollDirection)
    return true
  }

  private fun showOrHideView(child: View, @ScrollDirection direction: Int) {
    val hide = direction == VerticalScrollingBehavior.SCROLL_DIRECTION_UP
    if (this.hide == hide) return

    this.hide = hide

    if (anim == null) {
      anim = ViewCompat.animate(child)
      anim?.duration = 300
      anim?.interpolator = LinearOutSlowInInterpolator()
    } else {
      anim?.cancel()
    }
    val scale = (if (hide) 0 else 1).toFloat()
    anim!!.scaleX(scale).scaleY(scale).start()
  }
}
